src/i_oalmusic.c

```diff
@@ -87,6 +92,127 @@ static SDL_atomic_t player_thread_running;
 
 static boolean music_initialized;
 
+static ebur128_state *ebur_state;
+
+static void InitAutoGain(void)
+{
+    if (player.format == AL_FORMAT_MONO16 || player.format == AL_FORMAT_MONO_FLOAT32)
+    {
+        player.channels = 1;
+    }
+    else
+    {
+        player.channels = 2;
+    }
+
+    ebur_state = ebur128_init(player.channels, player.freq, EBUR128_MODE_S
+        | EBUR128_MODE_I | EBUR128_MODE_LRA | EBUR128_MODE_SAMPLE_PEAK
+        | EBUR128_MODE_HISTOGRAM);
+
+    player.auto_gain = 1.0f;
+}
+
+static void ShutdownAutoGain(void)
+{
+    if (ebur_state)
+    {
+        ebur128_destroy(&ebur_state);
+    }
+}
+
+#define LINEAR_TO_DB(amp) ((amp) >= 0.00001 ? 20.0 * log10((amp)) : -99.0)
+
+static void AutoGain(uint32_t frames)
+{
+    if (player.format == AL_FORMAT_MONO16 || player.format == AL_FORMAT_STEREO16)
+    {
+        ebur128_add_frames_short(ebur_state, (int16_t *)player.data, frames);
+    }
+    else
+    {
+        ebur128_add_frames_float(ebur_state, (float *)player.data, frames);
+    }
+
+    boolean failed = false;
+    double momentary = 0.0;
+    double shortterm = 0.0;
+    double global = 0.0;
+    double relative = 0.0;
+    double range = 0.0;
+
+    if (EBUR128_SUCCESS != ebur128_loudness_momentary(ebur_state, &momentary))
+    {
+        failed = true;
+    }
+
+    if (EBUR128_SUCCESS != ebur128_loudness_shortterm(ebur_state, &shortterm))
+    {
+        failed = true;
+    }
+
+    if (EBUR128_SUCCESS != ebur128_loudness_global(ebur_state, &global))
+    {
+        failed = true;
+    }
+
+    if (EBUR128_SUCCESS != ebur128_relative_threshold(ebur_state, &relative))
+    {
+        failed = true;
+    }
+
+    if (EBUR128_SUCCESS != ebur128_loudness_range(ebur_state, &range))
+    {
+        failed = true;
+    }
+
+    if (relative > -70.0 && momentary > relative && !failed)
+    {
+        double peak_L = 0.0;
+        double peak_R = 0.0;
+
+        if (EBUR128_SUCCESS != ebur128_prev_sample_peak(ebur_state, 0, &peak_L))
+        {
+            failed = true;
+        }
+
+        if (player.channels == 2)
+        {
+            if (EBUR128_SUCCESS
+                != ebur128_prev_sample_peak(ebur_state, 1, &peak_R))
+            {
+                failed = true;
+            }
+        }
+
+        if (!failed)
+        {
+            //float loudness = cbrtf(momentary * shortterm * global);
+            const float weight_m = 0.1f;
+            const float weight_s = 1.0f;
+            const float weight_i = 4.0f;
+            float loudness = (weight_m * momentary + weight_s * shortterm
+                              + weight_i * global)
+                             / (weight_m + weight_s + weight_i);
```

Comment:
@ceski-1  With `weight_m/weight_s/weight_i` we can adjust the weight of long term and short term methods. This will reduce gain deeps, try to tweak it to your liking.

Reply:
Okay, when I'm free again I'll take a closer look.